File: tasks/expr-eval/typescript/solution/evaluate.ts
import { Node, Program, ProgramError } from "./parse";

function wrap(value: bigint): bigint {
  return BigInt.asIntN(64, value);
}

function flooredQuotient(left: bigint, right: bigint): bigint {
  let quotient = left / right;
  if (left % right !== 0n && left < 0n !== right < 0n) {
    quotient -= 1n;
  }
  return quotient;
}

function divide(left: bigint, right: bigint, at: number): bigint {
  if (right === 0n) {
    throw new ProgramError("DIVIDE_BY_ZERO", at);
  }
  if (process.env["LAB_SABOTAGE"] === "truncate-toward-negative") {
    return wrap(flooredQuotient(left, right));
  }
  return wrap(left / right);
}

function remainder(left: bigint, right: bigint, at: number): bigint {
  if (right === 0n) {
    throw new ProgramError("DIVIDE_BY_ZERO", at);
  }
  if (process.env["LAB_SABOTAGE"] === "truncate-toward-negative") {
    return wrap(left - flooredQuotient(left, right) * right);
  }
  return wrap(left % right);
}

function shift(operator: string, left: bigint, right: bigint, at: number): bigint {
  let count = right;
  if (count < 0n || count > 63n) {
    if (process.env["LAB_SABOTAGE"] !== "shift-count-unchecked") {
      throw new ProgramError("SHIFT_RANGE", at);
    }
    count &= 63n;
  }
  return wrap(operator === "<<" ? left << count : left >> count);
}

function applyOperator(operator: string, left: bigint, right: bigint, at: number): bigint {
  switch (operator) {
    case "+":
      return wrap(left + right);
    case "-":
      return wrap(left - right);
    case "*":
      return wrap(left * right);
    case "/":
      return divide(left, right, at);
    case "%":
      return remainder(left, right, at);
    case "<<":
    case ">>":
      return shift(operator, left, right, at);
    case "&":
      return wrap(left & right);
    case "|":
      return wrap(left | right);
    case "^":
      return wrap(left ^ right);
    case "==":
      return left === right ? 1n : 0n;
    case "!=":
      return left !== right ? 1n : 0n;
    case "<":
      return left < right ? 1n : 0n;
    case "<=":
      return left <= right ? 1n : 0n;
    case ">":
      return left > right ? 1n : 0n;
    default:
      return left >= right ? 1n : 0n;
  }
}

function evaluateNode(node: Node, scope: Map<string, bigint>): bigint {
  if (node.kind === "literal") {
    return wrap(node.value);
  }
  if (node.kind === "name") {
    const value = scope.get(node.name);
    if (value === undefined) {
      throw new ProgramError("UNDEFINED", node.at);
    }
    return value;
  }
  if (node.kind === "unary") {
    const value = evaluateNode(node.operand, scope);
    return wrap(node.operator === "-" ? -value : ~value);
  }
  const left = evaluateNode(node.left, scope);
  const right = evaluateNode(node.right, scope);
  return applyOperator(node.operator, left, right, node.at);
}

export function evaluate(program: Program): bigint {
  const scope = new Map<string, bigint>();
  for (const binding of program.bindings) {
    scope.set(binding.name, evaluateNode(binding.node, scope));
  }
  return evaluateNode(program.body, scope);
}

File: tasks/expr-eval/go/solution/evaluate.go
package main

import "os"

func flooredQuotient(left int64, right int64) int64 {
	quotient := left / right
	if left%right != 0 && (left < 0) != (right < 0) {
		quotient--
	}
	return quotient
}

func divide(left int64, right int64, at int) (int64, error) {
	if right == 0 {
		return 0, &ProgramError{Code: "DIVIDE_BY_ZERO", At: at}
	}
	if os.Getenv("LAB_SABOTAGE") == "truncate-toward-negative" {
		return flooredQuotient(left, right), nil
	}
	return left / right, nil
}

func remainder(left int64, right int64, at int) (int64, error) {
	if right == 0 {
		return 0, &ProgramError{Code: "DIVIDE_BY_ZERO", At: at}
	}
	if os.Getenv("LAB_SABOTAGE") == "truncate-toward-negative" {
		return left - flooredQuotient(left, right)*right, nil
	}
	return left % right, nil
}

func shift(operator string, left int64, right int64, at int) (int64, error) {
	count := right
	if count < 0 || count > 63 {
		if os.Getenv("LAB_SABOTAGE") != "shift-count-unchecked" {
			return 0, &ProgramError{Code: "SHIFT_RANGE", At: at}
		}
		count &= 63
	}
	if operator == "<<" {
		return left << uint64(count), nil
	}
	return left >> uint64(count), nil
}

func boolValue(value bool) int64 {
	if value {
		return 1
	}
	return 0
}

func applyOperator(operator string, left int64, right int64, at int) (int64, error) {
	switch operator {
	case "+":
		return left + right, nil
	case "-":
		return left - right, nil
	case "*":
		return left * right, nil
	case "/":
		return divide(left, right, at)
	case "%":
		return remainder(left, right, at)
	case "<<", ">>":
		return shift(operator, left, right, at)
	case "&":
		return left & right, nil
	case "|":
		return left | right, nil
	case "^":
		return left ^ right, nil
	case "==":
		return boolValue(left == right), nil
	case "!=":
		return boolValue(left != right), nil
	case "<":
		return boolValue(left < right), nil
	case "<=":
		return boolValue(left <= right), nil
	case ">":
		return boolValue(left > right), nil
	default:
		return boolValue(left >= right), nil
	}
}

func evaluateNode(current *node, scope map[string]int64) (int64, error) {
	switch current.Kind {
	case "literal":
		return current.Value, nil
	case "name":
		value, ok := scope[current.Name]
		if !ok {
			return 0, &ProgramError{Code: "UNDEFINED", At: current.At}
		}
		return value, nil
	case "unary":
		value, err := evaluateNode(current.Left, scope)
		if err != nil {
			return 0, err
		}
		if current.Operator == "-" {
			return -value, nil
		}
		return ^value, nil
	}
	left, err := evaluateNode(current.Left, scope)
	if err != nil {
		return 0, err
	}
	right, err := evaluateNode(current.Right, scope)
	if err != nil {
		return 0, err
	}
	return applyOperator(current.Operator, left, right, current.At)
}

func evaluate(parsed *program) (int64, error) {
	scope := map[string]int64{}
	for _, item := range parsed.Bindings {
		value, err := evaluateNode(item.Node, scope)
		if err != nil {
			return 0, err
		}
		scope[item.Name] = value
	}
	return evaluateNode(parsed.Body, scope)
}
